core/iwasm/include/wasm_export.h

```diff
@@ -864,6 +880,34 @@ wasm_runtime_create_exec_env(wasm_module_inst_t module_inst,
 WASM_RUNTIME_API_EXTERN void
 wasm_runtime_destroy_exec_env(wasm_exec_env_t exec_env);
 
+
+/**
+ * @brief Copy callstack frames.
+ *
+ * Caution: This is not a thread-safe function. Ensure the exec_env
+ * is suspended before calling it from another thread.
+ *
+ * Usage: In the callback to read frames fields use APIs
+ * for wasm_frame_t from wasm_c_api.h
+ *
+ * Note: The function is async-signal-safe if called with verified arguments.
+ * Meaning it's safe to call it from a signal handler even on a signal
+ * interruption from another thread if next variables hold valid pointers
+ * - exec_env
+ * - exec_env->module_inst
+ * - exec_env->module_inst->module
+ *
+ * @param exec_env the execution environment that containes frames
+ * @param buffer the buffer of size equal length * sizeof(frame) to copy frames to
+ * @param length the number of frames to copy
+ * @param skip_n the number of frames to skip from the top of the stack
+ *
+ * @return number of copied frames
+ */
+WASM_RUNTIME_API_EXTERN uint32_t
+wasm_copy_callstack(const wasm_exec_env_t exec_env, wasm_frame_ptr_t buffer,
+                       const uint32_t length, const uint32_t skip_n);
+
```

Comment:
Are `length` and `skin_n` intended to track the resources consumed by buffer, such as copying the first half of the call stack at one time and the remaining half at another time? If that's the case, wouldn't it be quite challenging to prevent the call stack from changing between two consecutive copy operations?

Reply:
> such as copying the first half of the call stack at one time and the remaining half at another time?

Yes, the intention is to be able to be able to copy entire callstack in batches even if memory where it's being copied is limited

> Wouldn't it be quite challenging to prevent the call stack from changing between two consecutive copy operations?

If thread is paused/suspended/interrupted then it's all good. It's the case for us and should be the case for any user of this API in its current state anyway